File: test/pattern/test_facade.cpp
#include <gtest/gtest.h>
#include "pattern/facade.hpp"

using namespace pattern;

// ---------------------------------------------------------------------------
// Subsystem tests (direct use)
// ---------------------------------------------------------------------------
TEST(Facade, VideoFileCodecMp4)
{
    VideoFile f("movie.mp4");
    EXPECT_EQ("h264", f.codec());
}

TEST(Facade, VideoFileCodecMkv)
{
    VideoFile f("show.mkv");
    EXPECT_EQ("vp9", f.codec());
}

TEST(Facade, VideoFileCodecUnknown)
{
    VideoFile f("clip.avi");
    EXPECT_EQ("unknown", f.codec());
}

TEST(Facade, CodecFactoryExtractContainsCodecName)
{
    VideoFile f("film.mp4");
    std::string result = CodecFactory::extract(f);
    EXPECT_NE(std::string::npos, result.find("h264"));
}

TEST(Facade, BitrateReaderReadContainsPath)
{
    VideoFile f("sample.mp4");
    std::string data = BitrateReader::read(f, "h264");
    EXPECT_NE(std::string::npos, data.find("sample.mp4"));
}

TEST(Facade, BitrateReaderConvertContainsOriginalData)
{
    std::string conv = BitrateReader::convert("RAW_DATA", "vp9");
    EXPECT_NE(std::string::npos, conv.find("RAW_DATA"));
}

TEST(Facade, AudioMixerFixContainsInput)
{
    std::string mixed = AudioMixer::fix("VIDEO_DATA");
    EXPECT_NE(std::string::npos, mixed.find("VIDEO_DATA"));
}

// ---------------------------------------------------------------------------
// Facade tests
// ---------------------------------------------------------------------------
TEST(Facade, ConvertReturnsNonEmptyString)
{
    VideoConverter vc;
    std::string result = vc.convert("input.mp4", "mkv");
    EXPECT_FALSE(result.empty());
}

TEST(Facade, ConvertStoresLastResult)
{
    VideoConverter vc;
    std::string result = vc.convert("clip.mkv", "mp4");
    EXPECT_EQ(result, vc.lastResult());
}

TEST(Facade, ConvertContainsMixedAudio)
{
    VideoConverter vc;
    std::string result = vc.convert("video.mp4", "mkv");
    EXPECT_NE(std::string::npos, result.find("MixedAudio"));
}

TEST(Facade, ConvertMultipleFiles)
{
    VideoConverter vc;
    auto r1 = vc.convert("a.mp4", "mkv");
    auto r2 = vc.convert("b.mkv", "mp4");
    EXPECT_NE(r1, r2);
    EXPECT_EQ(r2, vc.lastResult());
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: scripts/test_pattern/test_facade.py
"""Tests for the Facade pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from facade import VideoFile, CodecFactory, BitrateReader, AudioMixer, VideoConverter


class TestFacade(unittest.TestCase):

    # ------------------------------------------------------------------
    # Subsystem (direct use)
    # ------------------------------------------------------------------
    def test_video_file_codec_mp4(self):
        self.assertEqual("h264", VideoFile("movie.mp4").codec())

    def test_video_file_codec_mkv(self):
        self.assertEqual("vp9", VideoFile("show.mkv").codec())

    def test_codec_factory_contains_codec(self):
        result = CodecFactory.extract(VideoFile("film.mp4"))
        self.assertIn("h264", result)

    def test_bitrate_reader_contains_path(self):
        data = BitrateReader.read(VideoFile("sample.mp4"), "h264")
        self.assertIn("sample.mp4", data)

    def test_bitrate_reader_convert_contains_data(self):
        conv = BitrateReader.convert("RAW_DATA", "vp9")
        self.assertIn("RAW_DATA", conv)

    def test_audio_mixer_contains_input(self):
        self.assertIn("VIDEO_DATA", AudioMixer.fix("VIDEO_DATA"))

    # ------------------------------------------------------------------
    # Facade
    # ------------------------------------------------------------------
    def test_convert_returns_non_empty(self):
        self.assertTrue(len(VideoConverter().convert("input.mp4", "mkv")) > 0)

    def test_convert_stores_last_result(self):
        vc = VideoConverter()
        result = vc.convert("clip.mkv", "mp4")
        self.assertEqual(result, vc.last_result)

    def test_convert_contains_mixed_audio(self):
        self.assertIn("MixedAudio", VideoConverter().convert("video.mp4", "mkv"))

    def test_convert_multiple_files_different_results(self):
        vc = VideoConverter()
        r1 = vc.convert("a.mp4", "mkv")
        r2 = vc.convert("b.mkv", "mp4")
        self.assertNotEqual(r1, r2)
        self.assertEqual(r2, vc.last_result)


if __name__ == "__main__":
    unittest.main()
